Microsoft Teams deltas and added features

billbliss opened the pull request:
Let me know how I might factor this differently.

msftclas commented:
@billbliss,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

billbliss commented:
I didn't mention the rationale for the edit on line 55. I deleted the "cb" parameter because it was not referenced anywhere in the code.

maxpert commented on src/adapter.coffee:
```diff
@@ -41,48 +40,105 @@ BotBuilder = require 'botbuilder'
+            # Retrieve Microsoft Teams tenant information if present to persist in the brain
+            # If not running inside Teams, tenant will be null
+            if activity.sourceEvent?
+                tenant = if activity.sourceEvent.tenant? then activity.sourceEvent.tenant.id else null
```
I believe BotFrameworkAdapter instead of inspecting stuff and creating variables that are not related to BF, it can have a list of middlewares. Given the source we can go through the list of middlewares and check if any of middlewares can handle the activity. If yes, it hands off the message creation task to middleware (the new TextMessage that we are doing right now on line 70-73 will be replaced), and then simply do ```@robot.receive message```. 

 Right now there is too much knowledge leaked into Adapter and that makes me uncomfortable.

billbliss replied:
Both BotBuilder.ChatConnector and Hubot can have middleware. I agree that's a clean way of doing it in theory, but it'll take a little bit of thought because there are some weird interactions between the Hubot layer and the BotBuilder layer.  For example, you'd have to pass the Hubot User and @robot objects into the middleware to run the above code. 

And wrt to 70-73, how is BotBuilder going to create a new Hubot TextMessage object?

Might be a real PITA to debug or maybe it's simple but I'm not sure I have my head around it.

maxpert commented on src/adapter.coffee:
```diff
@@ -41,48 +40,105 @@ BotBuilder = require 'botbuilder'
                 type: 'message'
                 text: str
-                address: context.user.activity.address
+                attachments: [
```
I think over here we need to support different object types that send can handle. Right now we are assuming all of them to be strings. So instead of us making these objects and choosing what goes over wire we can hand it off to the SDK consumer. This not only makes it powerful but also allows us to keep logic cleaner here.

maxpert commented on src/adapter.coffee:
```diff
@@ -41,48 +40,105 @@ BotBuilder = require 'botbuilder'
+            if imageAttachment isnt null and str is imageAttachment.contentUrl
+              msg.text = null
+            if imageAttachment isnt null 
+              msg.attachments.push(imageAttachment)
```
Again once we start understanding different types of messages here, we can obviously get rid of this code

maxpert commented on src/adapter.coffee:
```diff
@@ -41,48 +40,105 @@ BotBuilder = require 'botbuilder'
+# Helper functions for Bot Framework / Microsoft Teams
+
+# Transform Bot Framework/Microsoft Teams @mentions into Hubot's name as configured
+hubotifyBotMentions = (msgText, mentions, bfBotId, hubotBotName) ->
```
Again I think this can go to middleware.

maxpert commented:
I made some changes over the weekend for middle-ware. I still need to test them but so far I can see lots of channel specific parts can be moved to middle-wares. I am not sure however on if it's a good idea to have an adapter that can further have middle-wares. It's like saying adapter having adapters 😛

billbliss commented:
Zohaib, I have integrated my branch with your changes.

- adapter.coffee should be exactly the same as your most recent version. But since there was a merge conflict, I had to submit a new version of the file.
- msteams-middleware.coffee has not changed and works the same way as it did before and plugs into Hubot middleware. 
- middlewares.coffee incorporates all the changes I couldn't originally factor out of adapter.coffee. There are three significant deltas from your version:
  - Adding the Microsoft Teams tenant id to the robot.brain entry in TextMiddleware. Without replacing your middleware with a completely different version I wasn't sure how to factor this. That is, once your middleware writes the entry, I can't add another middleware to write another one. The change should  not affect other channels.
  - Changing the @mentions to match Teams semantics. This one might be factor-able, but I wasn't sure.
  - Handling the special case of when the text response contains a URL - adding it to the attachments array. I think that's generally useful across all the BF channels.
  - There are a few helper functions I added. If there's a cleaner way of adding those, let me know.

billbliss commented:
Zohaib, 

I spent the holiday weekend making the changes you suggested as well as a few other things to clean it up:
- Renamed LICENSE to LICENSE.md which is the standard naming convention for license documents. Also moved the redundant MIT license from the source and added "Licensed under the MIT license" comments, following CELA OSS standards.
- Fixed a typo in package.json
- Renamed "middlewares.coffee" to "adapter-middleware.coffee" to distinguish between Hubot middleware and this. Also renamed `middlewares` to `middleware` (proper English)
- Reduced the imports from adapter-middleware (formerly middlewares) - only the `middlewareFor` method is needed
- Added a comment and a commented-out line to make it easy to activate MS Teams middleware. I tried doing an include of it instead as a script, as you suggested, but then you get an error during the build process because the Hubot script loader expects functions, not class definitions. This is a non-invasive change IMO.
- Fixed the middleware invocation logic in adapter.coffee - you were using `channelId` but it needs to be `activity.source`
- Minor change to error handling syntax in `reply:` method in adapter.coffee to make it easier to add real error handling later
- msteams-middleware.coffee is structured as you suggested - thanks. I haven't refactored the Hubot middleware into toSendable middleware though because I'm having some problems (unrelated) with being throttled when submitting multiple messages, so I want to get to the bottom of that first.

MattSFT commented on src/msteams-middleware.coffee:
```diff
@@ -0,0 +1,181 @@
+    # Initialize Hubot Middleware
+
+    # Sends a typing indicator before sending a message
+    @robot.responseMiddleware (context, next, done) ->
```
It should really send the typing indicator when it receives the messages, right? I thought it was supposed to be an indicator of the bot thinking

billbliss replied:
As opposed to when it *sends* the first one, that is? That's a good point. That might be a better time to do it. Let me think about that a bit.

MattSFT commented on src/msteams-middleware.coffee:
```diff
@@ -0,0 +1,181 @@
+    if activity.sourceEvent?
+      tenant = if activity.sourceEvent.tenant? then activity.sourceEvent.tenant.id else null
+
+    user = @robot.brain.userForId address.user.id, name: address.user.name, room: address.conversation.id, tenant: tenant
```
I heard that user Ids are not guaranteed to be unique across tenants so a malicious user may try to set their id to one of another user in the system. Maybe the id we store the user under is a concatenation #{tenantId}_#{userId} if tenant id is set?

billbliss commented:
Merge conflicts resolved.

msftclas commented:
@billbliss,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_
